### ISSUE TITLE
Wherigo player: Some findings

### ISSUE THREAD

**Issue by moving-bits** [cgeo/cgeo#16279]
I played a bit with PR #16276 - really like the new Wherigo player, well integrated into c:geo! Thanks for putting all the effort in it.

Some small findings: 

(1) When activating "Wherigo player" (eg: from system notification) it highlights the fifth element in bottom navigation, even if it's still configured as "nearby".

(2) Configured the last item in bottom navigation as "none" and tapped on system navigation. Wherigo activity opened, but "back" arrow in title bar does not work (does nothing). (Android's back tap returns to c:geo home screen, which would be expected behavior for back arrow as well.)

(3) BTW: Somehow the "paused dialog" got lost during those tests. I found no way to resume it. (Would have expected that tapping on the system notification would bring me back.) Screen looks like this:

![image](https://github.com/user-attachments/assets/c7147dd7-2473-49e3-956c-043a1dd87712)

"Play" button wants to load a new cartridge, loading a savegame would stop the running game. :man_shrugging: 

(4) Ok, "loading a savegame" crashed c:geo:

```
--------- beginning of crash
2024-10-27 22:59:55.230  4307-4307  AndroidRuntime          cgeo.geocaching.developer            E  FATAL EXCEPTION: main
    Process: cgeo.geocaching.developer, PID: 4307
    io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.util.ConcurrentModificationException
     at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:372)
     at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:125)
     at android.os.Handler.handleCallback(Handler.java:938)
     at android.os.Handler.dispatchMessage(Handler.java:99)
     at android.os.Looper.loop(Looper.java:223)
     at android.app.ActivityThread.main(ActivityThread.java:7656)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
    Caused by: java.util.ConcurrentModificationException
     at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
     at java.util.HashMap$ValueIterator.next(HashMap.java:1470)
     at cgeo.geocaching.wherigo.WherigoGame.lambda$notifyListeners$2$cgeo-geocaching-wherigo-WherigoGame(WherigoGame.java:240)
     at cgeo.geocaching.wherigo.WherigoGame$$ExternalSyntheticLambda4.run(Unknown Source:4)
     at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:123)
     at android.os.Handler.handleCallback(Handler.java:938) 
     at android.os.Handler.dispatchMessage(Handler.java:99) 
     at android.os.Looper.loop(Looper.java:223) 
     at android.app.ActivityThread.main(ActivityThread.java:7656) 
     at java.lang.reflect.Method.invoke(Native Method) 
     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
2024-10-27 22:59:55.643  7094-7094  cgeo                    cgeo.geocaching.developer            E  [main] c:geo version 2024.10.27-133cc39 developer build
2024-10-27 22:59:55.733  7094-7094  ching.develope          cgeo.geocaching.developer            E  Invalid ID 0x00000000.
2024-10-27 22:59:57.670  7094-7142  Bitmap                  cgeo.geocaching.developer            A  Error, cannot access an invalid/free'd bitmap here!
2024-10-27 22:59:57.875  7094-7142  ching.develope          cgeo.geocaching.developer            A  runtime.cc:655] Runtime aborting...
    runtime.cc:655] Dumping all threads without mutator lock held
```
(GitHub will not accept the full log due to max. size limit of 64k - I will send the full log to support account)

Then I restarted c:geo, tapped on the WIG icon, cache details, play WIG - and got dialog where I could load the autosaved game :+1:

I was able to re-enter the zone and to read the text (either by using the WIG menu => locations or by tapping on the red-lined area on the map), the description of the zone got displayed, but the dialog with the question for this zone did not show up again. Guess I will have to restart. (Tomorrow maybe, as time permits)

After step 2 above, the bottom navigation did not show up anymore while being in Wherigo mode, even when coming from the map by clicking into a zone. Not sure if this is the intended behavior.

(5) Update: Another finding: Started c:geo in a different emulator, loaded the mentioned cartridge via "search by gc code", opened cache details and pressed the "play" button. It showed me the first location. For some reason I tapped on the system notification for the wherigo player, which brought me to the wherigo activity. Here I tapped on "Locations (1)", which brought up the listing for the first location. Tapping on "Show on map" just closed this view, returning to the wherigo activity instead of opening the map at the location specified by the zone. [Reason may be, that c:geo's "GPS" was defaulting to Paris at that point in time. If it's not allowed to show the zone on the map in that state, an error message should be displayed. IMHO.]

(6) Tapping on back arrow in the title bar had no impact here.

Played cartridge is GC93DPE "Alles zu seiner Zeit Bärlin"
on an emulated API 30 Pixel 3a device.

**Comment by eddiemuc** [comment]
So many bugs....
Thanks for the extensive test! Might be a few days before I can dive into this though.

**Comment by eddiemuc** [comment]
Wrt to (1) and (2): those are related to navigation bar and custom items in it. I am not really an expert on this and tried to integrate wherigo activity as best as it seem fit, trying to follow other activities being integrated there. I am not sure I did this right. Can someone with more knowledge in this area please take a look into this as well? @moving-bits ? @fm-sys ?

**Comment by moving-bits** [comment]
After some more playing around with today's master (which includes the mentioned PR), I'm no longer able to reproduce (1) - mostly it opens without a bottom navigation, and in the rare occasions where it showed a bottom navigation, the "map" button was highlighted (ok) - although I don't know how that worked... (I have an idea what would need to be changed to set the fifth button in navigation bar, if (and only if) this one is configured as Whereigo).

Maybe you can elaborate a bit when Wherigo player is supposed to show a bottom navigation and when not?

Wrt to (2): This seems to be independent from custom button navigation, see finding (6), which is a follow-up of (5), both added today.

**Comment by eddiemuc** [comment]
> Maybe you can elaborate a bit when Wherigo player is supposed to show a bottom navigation and when not?

That's simple and not simple at the same time: I want the Wherigo Activity ton behave exactly as any other c:geo Activity which can optionally be placed on the fifth custom slot. I didn't found any documentation about this, so I observed the behavior of Settings (can also be set on fifth slot). I observed and concluded:
- if an item is NOT part of bottom navigation then it should also never open with bottom navigation. 
- if it is at bottom nav and reached by clicking the bottom nav then bottom nav shall appear/stay
- a bit unclear: if we are in an activity with bottom nav but the item is selected by means other than the bottom nav (eg button, menu item), shall the bottom nav stay or not? (in case of Settings as custom item, the nav bar will go away when selecting Settings from menu in main activity. This felt unexpected to me but might have a deeper reason?)
- if we jump to the item from within an activity without a bottom nav (by means like buttons, menu, tap,...) then the item too shall not have a bottom nav

I am not sure whether I interpreted the wishes of the bottom nav inventors correctly, nor can I guarantee that I implemented it correctly. I feel however that this is not a Wherigo discussion but a bottom navigation discussion. If there are unclarities, shall we open a new issue for it?

**Comment by eddiemuc** [comment]
PR #16301 adresses several of the issues from above:
* Navigation Bar related thingies ((1) and (2) and mentioned in some other points as well lime (5)): PR implements its own behaviour for when to show/hide nav bar item. I honestly don't approve that this shall be a per-activity-thingy, and I can just hope I thought of all cases. For an overall cleanup of this issue I would like to refer to #16287 
* (4): crash on saving a game should be fixed
* (6): should be fixed, back arrow should work with the PR

**Comment by eddiemuc** [comment]
Wrt the map display problems mentioned in (5): @moving-bits : displaying wherigo zones in map should work also if GPS is not working / pointing to Paris (except if Unified Map has a problem starting up in such a situation?). I basically just call a Unified Map Start activity here. If this is not working: do you happen to have a logfile? There might be a hidden crash here.

**Comment by eddiemuc** [comment]
Wrt (3): "Somehow the "paused dialog" got lost during those tests.": this is a bit more complex and not easy to solve.

A bit of context on how Wherigos work: basically the Wherigo engine fires asynchron requests upon its "hosting application" to display dialogs. Those may be things which can also be selected manually by user (eg showing a "Thing", "Zone", "Item" and such), but those may also be things that which the user cannot navigate to later. Those are "Show Dialogs" and "Input Dialogs". The later two are the problematic ones because once they are dismissed again the user has no chance to re-open/re-navigate to them. Moreover, Wherigo-logic enforces the host application to dismiss those dialogs when another Wherigo dialog should be shown - regardless on whether the dialog was finished by the user.

Consequently, the "pause" function in c:geo can't "hold" such a dialog indefinitely. It will be dismissed in following cases:
* whenever wherigo player forces to show another dialog (this is necessary to comply with wherigo logic and thus not provoke unexpected game failures)
* when c:geo is stopped, either when user stops it or it is stopped/restarted by Android in background

This problem is not specific to c:geo's player, it is also a problem with other players (eg WhereYouGo): once you dismissed such a dialog you can't come back to it. Most cartridge programmers are aware of this and find ways around it. Most commonly they allow retriggering such dialogs via actions in a Zone or other item.

The cartridge in GC93DPE "Alles zu seiner Zeit Bärlin" above does not provide such a feature unfortunately. This is why by dismissing a dialog it is possible to get in a game state where there is no way out. Unfortunate but true. You can provoke the same in other players with this cartridge by dismissing eg the first input dialog.

**Comment by eddiemuc** [comment]
I hope I covered all things mentioned in this issue. Please give me a ping if I missed something

**Comment by moving-bits** [comment]
I've played around a bit with the new PR and was no longer able to provoke the issues mentioned above, so let's close here for now and open separate issues if something new comes up. Thanks for deep-diving into those findings!

**Comment by moving-bits** [comment]
@eddiemuc 
BTW - have you seen this article by Saarfuchs about insights into Wherigo cartrdiges?
https://www.saarfuchs.com/2024/10/wherigo-moeglichkeiten-zum-cheaten-und-ein-schutz-davor.html
